File: tests/fixtures/python_project/edge_cases.py
"""엣지 케이스 및 이상한 패턴들"""

import asyncio
from collections.abc import Callable
from contextlib import contextmanager
from enum import Enum, auto
from functools import wraps
from typing import (
    Any,
    Generic,
    Literal,
    Protocol,
    TypedDict,
    TypeVar,
    Union,
    overload,
    runtime_checkable,
)

# 복잡한 타입 별칭
T = TypeVar('T')
K = TypeVar('K')
V = TypeVar('V')

NestedDict = dict[str, str | int | dict[str, Any] | list[Any]]
ComplexType = Union[list[dict[str, int | str | None]], dict[str, list[int]]]


# TypedDict 사용
class UserData(TypedDict, total=False):
    name: str
    age: int
    email: str | None
    metadata: dict[str, Any]


# 복잡한 Enum
class Status(Enum):
    PENDING = auto()
    PROCESSING = auto()
    COMPLETED = auto()
    FAILED = auto()
    CANCELLED = auto()

    @classmethod
    def from_string(cls, value: str) -> "Status":
        """문자열로부터 Status 생성"""
        try:
            return cls[value.upper()]
        except KeyError:
            return cls.PENDING


# 메타클래스
class SingletonMeta(type):
    """싱글톤 메타클래스"""
    _instances: dict[type, Any] = {}

    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            cls._instances[cls] = super().__call__(*args, **kwargs)
        return cls._instances[cls]


class DatabaseConnection(metaclass=SingletonMeta):
    """싱글톤 데이터베이스 연결"""

    def __init__(self):
        self.connected = False

    def connect(self):
        self.connected = True

    def disconnect(self):
        self.connected = False


# 복잡한 제네릭 클래스
class Container(Generic[T]):
    """제네릭 컨테이너"""

    def __init__(self, value: T):
        self._value = value

    def get(self) -> T:
        return self._value

    def set(self, value: T) -> None:
        self._value = value

    def map(self, func: Callable[[T], K]) -> "Container[K]":
        return Container(func(self._value))


class MultiContainer(Generic[T, K]):
    """다중 타입 파라미터 제네릭"""

    def __init__(self, first: T, second: K):
        self.first = first
        self.second = second

    def swap(self) -> "MultiContainer[K, T]":
        return MultiContainer(self.second, self.first)


# 오버로드 사용
class Calculator:
    """오버로드를 사용하는 계산기"""

    @overload
    def add(self, x: int, y: int) -> int: ...

    @overload
    def add(self, x: str, y: str) -> str: ...

    @overload
    def add(self, x: list[int], y: list[int]) -> list[int]: ...

    def add(self, x: int | str | list[int], y: int | str | list[int]) -> int | str | list[int]:
        """타입에 따라 다른 덧셈 수행"""
        if isinstance(x, int) and isinstance(y, int) or isinstance(x, str) and isinstance(y, str):
            return x + y
        elif isinstance(x, list) and isinstance(y, list):
            return [a + b for a, b in zip(x, y, strict=False)]
        raise TypeError("지원하지 않는 타입")


# 복잡한 데코레이터 체이닝
def validate_input(validator: Callable[[Any], bool]):
    """입력 검증 데코레이터"""
    def decorator(func: Callable) -> Callable:
        @wraps(func)
        def wrapper(*args, **kwargs):
            for arg in args:
                if not validator(arg):
                    raise ValueError(f"Invalid input: {arg}")
            return func(*args, **kwargs)
        return wrapper
    return decorator


def log_execution(func: Callable) -> Callable:
    """실행 로깅 데코레이터"""
    @wraps(func)
    def wrapper(*args, **kwargs):
        print(f"Executing {func.__name__} with args={args}, kwargs={kwargs}")
        result = func(*args, **kwargs)
        print(f"Result: {result}")
        return result
    return wrapper


def retry_on_failure(max_retries: int = 3):
    """실패 시 재시도 데코레이터"""
    def decorator(func: Callable) -> Callable:
        @wraps(func)
        async def async_wrapper(*args, **kwargs):
            for attempt in range(max_retries):
                try:
                    return await func(*args, **kwargs)
                except Exception as e:
                    if attempt == max_retries - 1:
                        raise
                    print(f"Attempt {attempt + 1} failed: {e}")
            return None

        @wraps(func)
        def sync_wrapper(*args, **kwargs):
            for attempt in range(max_retries):
                try:
                    return func(*args, **kwargs)
                except Exception as e:
                    if attempt == max_retries - 1:
                        raise
                    print(f"Attempt {attempt + 1} failed: {e}")
            return None

        import asyncio
        if asyncio.iscoroutinefunction(func):
            return async_wrapper
        return sync_wrapper
    return decorator


@validate_input(lambda x: isinstance(x, int) and x > 0)
@log_execution
@retry_on_failure(max_retries=2)
def complex_function(value: int) -> int:
    """복잡한 데코레이터가 적용된 함수"""
    return value * 2


# 동적 속성을 가진 클래스
class DynamicClass:
    """동적으로 속성을 추가할 수 있는 클래스"""

    def __init__(self, **kwargs):
        for key, value in kwargs.items():
            setattr(self, key, value)

    def __getattr__(self, name: str) -> Any:
        """존재하지 않는 속성 접근 시 기본값 반환"""
        return f"<default_{name}>"

    def __setattr__(self, name: str, value: Any) -> None:
        """속성 설정 시 검증"""
        if name.startswith("_"):
            object.__setattr__(self, name, value)
        else:
            object.__setattr__(self, name, value)

    def __delattr__(self, name: str) -> None:
        """속성 삭제 시 특별 처리"""
        if name.startswith("protected_"):
            raise AttributeError(f"Cannot delete protected attribute: {name}")
        object.__delattr__(self, name)


# 복잡한 예외 처리
class CustomError(Exception):
    """커스텀 예외"""

    def __init__(self, message: str, code: int = 0):
        self.message = message
        self.code = code
        super().__init__(self.message)


class ValidationError(CustomError):
    """검증 에러"""
    pass


class ProcessingError(CustomError):
    """처리 에러"""
    pass


def risky_operation(data: Any) -> Any:
    """예외를 발생시킬 수 있는 작업"""
    if not data:
        raise ValidationError("Data is required", code=1001)

    if isinstance(data, str) and len(data) > 1000:
        raise ProcessingError("Data too large", code=2001)

    try:
        result = int(data) if isinstance(data, str) else data
        return result * 2
    except (ValueError, TypeError) as e:
        raise ProcessingError(f"Failed to process: {e}", code=2002) from e


# 복잡한 람다 및 함수형 패턴
def create_filter_function(threshold: int) -> Callable[[list[int]], list[int]]:
    """필터 함수를 생성하는 팩토리"""
    return lambda items: [x for x in items if x > threshold]


def compose(*functions: Callable) -> Callable:
    """함수 합성"""
    def composed(x: Any) -> Any:
        result = x
        for func in reversed(functions):
            result = func(result)
        return result
    return composed


# 복잡한 컨텍스트 매니저
@contextmanager
def managed_resource(name: str):
    """리소스 관리 컨텍스트 매니저"""
    print(f"Acquiring resource: {name}")
    resource = {"name": name, "acquired": True}
    try:
        yield resource
    finally:
        print(f"Releasing resource: {name}")
        resource["acquired"] = False


class ResourceManager:
    """리소스 관리자 클래스"""

    def __init__(self, name: str):
        self.name = name
        self.acquired = False

    def __enter__(self):
        self.acquired = True
        print(f"Entering context: {self.name}")
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.acquired = False
        print(f"Exiting context: {self.name}")
        return False  # 예외를 전파


# 복잡한 프로토콜
@runtime_checkable
class Drawable(Protocol):
    """그리기 가능한 프로토콜"""

    def draw(self) -> None: ...

    def get_area(self) -> float: ...


@runtime_checkable
class Movable(Protocol):
    """이동 가능한 프로토콜"""

    def move(self, x: float, y: float) -> None: ...


class Shape:
    """도형 클래스"""

    def __init__(self, x: float, y: float):
        self.x = x
        self.y = y

    def draw(self) -> None:
        print(f"Drawing shape at ({self.x}, {self.y})")

    def get_area(self) -> float:
        return 0.0


class Circle(Shape):
    """원 클래스"""

    def __init__(self, x: float, y: float, radius: float):
        super().__init__(x, y)
        self.radius = radius

    def get_area(self) -> float:
        return 3.14159 * self.radius ** 2

    def move(self, x: float, y: float) -> None:
        self.x += x
        self.y += y


# 복잡한 네스팅
class OuterClass:
    """외부 클래스"""

    class InnerClass:
        """내부 클래스"""

        class NestedClass:
            """중첩 클래스"""

            @staticmethod
            def deeply_nested_method() -> str:
                """깊게 중첩된 메서드"""
                return "deeply nested"

        def inner_method(self) -> str:
            """내부 메서드"""
            return "inner"

    def outer_method(self) -> "OuterClass.InnerClass":
        """외부 메서드"""
        return OuterClass.InnerClass()


# 이상한 네이밍 패턴
class _PrivateClass:
    """프라이빗 클래스 (언더스코어로 시작)"""
    pass


class __DunderClass__:
    """던더 클래스"""
    def __init__(self):
        self.__private_attr = "secret"


class CamelCaseClass:
    """카멜케이스 클래스"""
    def mixedCaseMethod(self) -> None:
        """혼합 케이스 메서드"""
        pass


class snake_case_class:
    """스네이크 케이스 클래스"""
    def method_name(self) -> None:
        pass


# 특수 메서드 오버로딩
class Vector:
    """벡터 클래스 (연산자 오버로딩)"""

    def __init__(self, x: float, y: float):
        self.x = x
        self.y = y

    def __add__(self, other: "Vector") -> "Vector":
        return Vector(self.x + other.x, self.y + other.y)

    def __sub__(self, other: "Vector") -> "Vector":
        return Vector(self.x - other.x, self.y - other.y)

    def __mul__(self, scalar: float) -> "Vector":
        return Vector(self.x * scalar, self.y * scalar)

    def __str__(self) -> str:
        return f"Vector({self.x}, {self.y})"

    def __repr__(self) -> str:
        return f"Vector(x={self.x}, y={self.y})"

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Vector):
            return False
        return self.x == other.x and self.y == other.y

    def __hash__(self) -> int:
        return hash((self.x, self.y))

    def __len__(self) -> int:
        return 2

    def __getitem__(self, index: int) -> float:
        if index == 0:
            return self.x
        elif index == 1:
            return self.y
        raise IndexError("Vector index out of range")

    def __iter__(self):
        yield self.x
        yield self.y


# 복잡한 비동기 패턴
class AsyncProcessor:
    """비동기 프로세서"""

    async def process_item(self, item: Any) -> Any:
        """아이템 처리"""
        await asyncio.sleep(0.1)
        return item * 2

    async def process_batch(self, items: list[Any]) -> list[Any]:
        """배치 처리"""
        tasks = [self.process_item(item) for item in items]
        return await asyncio.gather(*tasks)

    async def process_with_retry(self, item: Any, max_retries: int = 3) -> Any:
        """재시도와 함께 처리"""
        for attempt in range(max_retries):
            try:
                return await self.process_item(item)
            except Exception:
                if attempt == max_retries - 1:
                    raise
                await asyncio.sleep(0.1 * (attempt + 1))


# 데코레이터 팩토리
def register_handler(event_type: str):
    """핸들러 등록 데코레이터 팩토리"""
    handlers: dict[str, list[Callable]] = {}

    def decorator(func: Callable) -> Callable:
        if event_type not in handlers:
            handlers[event_type] = []
        handlers[event_type].append(func)
        return func

    decorator.handlers = handlers
    return decorator


@register_handler("user_created")
def handle_user_created(user: dict[str, Any]) -> None:
    """사용자 생성 핸들러"""
    print(f"User created: {user}")


@register_handler("user_deleted")
def handle_user_deleted(user_id: str) -> None:
    """사용자 삭제 핸들러"""
    print(f"User deleted: {user_id}")


# 복잡한 타입 가드
def is_valid_user(obj: Any) -> bool:
    """타입 가드 함수"""
    return (
        isinstance(obj, dict) and
        "name" in obj and
        "age" in obj and
        isinstance(obj["name"], str) and
        isinstance(obj["age"], int)
    )


def process_user_data(data: Any) -> dict[str, Any] | None:
    """타입 가드를 사용하는 함수"""
    if is_valid_user(data):
        return {"processed": True, "user": data}
    return None


# 이상한 주석 패턴
class CommentedClass:
    """
    매우 긴 독스트링을 가진 클래스.

    이 클래스는 여러 줄에 걸친 설명을 가지고 있으며,
    복잡한 기능을 수행합니다.

    Args:
        value: 초기값

    Returns:
        CommentedClass 인스턴스

    Raises:
        ValueError: 잘못된 값이 전달될 경우

    Example:
        >>> obj = CommentedClass(42)
        >>> print(obj.value)
        42
    """

    def __init__(self, value: int):
        # 인라인 주석
        self.value = value  # 값 설정

    def method(self):
        """메서드 독스트링"""
        # TODO: 구현 필요
        # FIXME: 버그 수정 필요
        # NOTE: 중요 참고사항
        # XXX: 위험한 코드
        pass


# 복잡한 조건부 타입 (타입 힌트)
def process_data(data: str | int | list[int]) -> str | int | list[int]:
    """조건부 처리를 하는 함수"""
    if isinstance(data, str):
        return data.upper()
    elif isinstance(data, int):
        return data * 2
    elif isinstance(data, list):
        return [x * 2 for x in data]
    else:
        raise TypeError("Unsupported type")


# Literal 타입 사용
def get_status_color(status: Literal["success", "error", "warning"]) -> str:
    """상태에 따른 색상 반환"""
    colors = {
        "success": "green",
        "error": "red",
        "warning": "yellow"
    }
    return colors[status]


# 복잡한 함수 시그니처
def complex_function_signature(
    required: str,
    optional: int | None = None,
    *args: str,
    keyword_only: bool = False,
    **kwargs: Any
) -> dict[str, Any]:
    """복잡한 함수 시그니처"""
    return {
        "required": required,
        "optional": optional,
        "args": args,
        "keyword_only": keyword_only,
        "kwargs": kwargs
    }

File: tests/fixtures/typescript_project/edge_cases.ts
/**
 * 엣지 케이스 및 이상한 패턴들
 */

// 복잡한 타입 정의
type NestedDict = {
  [key: string]: string | number | NestedDict | any[];
};

type ComplexType =
  | Array<{ [key: string]: number | string | null }>
  | { [key: string]: number[] };

type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends object ? DeepPartial<T[P]> : T[P];
};

type RequiredFields<T, K extends keyof T> = T & Required<Pick<T, K>>;

// 조건부 타입
type NonNullable<T> = T extends null | undefined ? never : T;

type FunctionReturnType<T> = T extends (...args: any[]) => infer R ? R : never;

type ArrayElementType<T> = T extends (infer U)[] ? U : never;

// 복잡한 제네릭
interface Container<T> {
  value: T;
  get(): T;
  set(value: T): void;
  map<U>(fn: (value: T) => U): Container<U>;
}

class GenericContainer<T> implements Container<T> {
  constructor(public value: T) {}

  get(): T {
    return this.value;
  }

  set(value: T): void {
    this.value = value;
  }

  map<U>(fn: (value: T) => U): Container<U> {
    return new GenericContainer(fn(this.value));
  }
}

interface MultiContainer<T, K> {
  first: T;
  second: K;
  swap(): MultiContainer<K, T>;
}

// 오버로드
class Calculator {
  add(x: number, y: number): number;
  add(x: string, y: string): string;
  add(x: number[], y: number[]): number[];
  add(
    x: number | string | number[],
    y: number | string | number[]
  ): number | string | number[] {
    if (typeof x === "number" && typeof y === "number") {
      return x + y;
    }
    if (typeof x === "string" && typeof y === "string") {
      return x + y;
    }
    if (Array.isArray(x) && Array.isArray(y)) {
      return x.map((val, idx) => val + y[idx]);
    }
    throw new Error("Unsupported types");
  }
}

// 싱글톤 패턴
class DatabaseConnection {
  private static instance: DatabaseConnection;
  private connected: boolean = false;

  private constructor() {}

  static getInstance(): DatabaseConnection {
    if (!DatabaseConnection.instance) {
      DatabaseConnection.instance = new DatabaseConnection();
    }
    return DatabaseConnection.instance;
  }

  connect(): void {
    this.connected = true;
  }

  disconnect(): void {
    this.connected = false;
  }

  isConnected(): boolean {
    return this.connected;
  }
}

// 복잡한 데코레이터 패턴 (TypeScript 데코레이터)
function validateInput(validator: (value: any) => boolean) {
  return function (
    target: any,
    propertyKey: string,
    descriptor: PropertyDescriptor
  ) {
    const originalMethod = descriptor.value;

    descriptor.value = function (...args: any[]) {
      for (const arg of args) {
        if (!validator(arg)) {
          throw new Error(`Invalid input: ${arg}`);
        }
      }
      return originalMethod.apply(this, args);
    };

    return descriptor;
  };
}

function logExecution(
  target: any,
  propertyKey: string,
  descriptor: PropertyDescriptor
) {
  const originalMethod = descriptor.value;

  descriptor.value = function (...args: any[]) {
    console.log(`Executing ${propertyKey} with args:`, args);
    const result = originalMethod.apply(this, args);
    console.log(`Result:`, result);
    return result;
  };

  return descriptor;
}

class ProcessedClass {
  @validateInput((x: any) => typeof x === "number" && x > 0)
  @logExecution
  process(value: number): number {
    return value * 2;
  }
}

// 동적 속성 클래스
class DynamicClass {
  [key: string]: any;

  constructor(initialData: Record<string, any> = {}) {
    Object.assign(this, initialData);
  }

  getProperty(name: string): any {
    return this[name] ?? `<default_${name}>`;
  }

  setProperty(name: string, value: any): void {
    if (!name.startsWith("_")) {
      this[name] = value;
    }
  }
}

// 복잡한 예외 처리
class CustomError extends Error {
  code: number;

  constructor(message: string, code: number = 0) {
    super(message);
    this.name = "CustomError";
    this.code = code;
  }
}

class ValidationError extends CustomError {
  constructor(message: string) {
    super(message, 1001);
    this.name = "ValidationError";
  }
}

class ProcessingError extends CustomError {
  constructor(message: string) {
    super(message, 2001);
    this.name = "ProcessingError";
  }
}

function riskyOperation(data: any): any {
  if (!data) {
    throw new ValidationError("Data is required");
  }

  if (typeof data === "string" && data.length > 1000) {
    throw new ProcessingError("Data too large");
  }

  try {
    const result = typeof data === "string" ? parseInt(data) : data;
    return result * 2;
  } catch (error) {
    throw new ProcessingError(`Failed to process: ${error}`);
  }
}

// 복잡한 함수형 패턴
type FilterFunction = (items: number[]) => number[];

function createFilterFunction(threshold: number): FilterFunction {
  return (items: number[]) => items.filter(x => x > threshold);
}

type ComposeFunction = <T>(...fns: Array<(x: T) => T>) => (x: T) => T;

const compose: ComposeFunction = <T>(...fns: Array<(x: T) => T>) => {
  return (x: T) => fns.reduceRight((acc, fn) => fn(acc), x);
};

// 복잡한 프로토콜/인터페이스
interface Drawable {
  draw(): void;
  getArea(): number;
}

interface Movable {
  move(x: number, y: number): void;
}

abstract class Shape implements Drawable {
  constructor(protected x: number, protected y: number) {}

  abstract getArea(): number;

  draw(): void {
    console.log(`Drawing shape at (${this.x}, ${this.y})`);
  }
}

class Circle extends Shape implements Movable {
  constructor(x: number, y: number, private radius: number) {
    super(x, y);
  }

  getArea(): number {
    return Math.PI * this.radius ** 2;
  }

  move(x: number, y: number): void {
    this.x += x;
    this.y += y;
  }
}

// 복잡한 네스팅
class OuterClass {
  static InnerClass = class {
    static NestedClass = class {
      static deeplyNestedMethod(): string {
        return "deeply nested";
      }
    };

    innerMethod(): string {
      return "inner";
    }
  };

  outerMethod(): OuterClass.InnerClass {
    return new OuterClass.InnerClass();
  }
}

// 이상한 네이밍 패턴
class _PrivateClass {
  private __privateAttr: string = "secret";
}

class __DunderClass__ {
  private __private_attr: string = "secret";
}

class CamelCaseClass {
  mixedCaseMethod(): void {}
}

class snake_case_class {
  method_name(): void {}
}

// 연산자 오버로딩 (TypeScript에서는 직접 지원하지 않지만 패턴으로 구현)
class Vector {
  constructor(public x: number, public y: number) {}

  add(other: Vector): Vector {
    return new Vector(this.x + other.x, this.y + other.y);
  }

  subtract(other: Vector): Vector {
    return new Vector(this.x - other.x, this.y - other.y);
  }

  multiply(scalar: number): Vector {
    return new Vector(this.x * scalar, this.y * scalar);
  }

  toString(): string {
    return `Vector(${this.x}, ${this.y})`;
  }

  equals(other: Vector): boolean {
    return this.x === other.x && this.y === other.y;
  }

  get length(): number {
    return Math.sqrt(this.x ** 2 + this.y ** 2);
  }

  [Symbol.iterator](): Iterator<number> {
    let index = 0;
    const values = [this.x, this.y];
    return {
      next(): IteratorResult<number> {
        if (index < values.length) {
          return { value: values[index++], done: false };
        }
        return { value: undefined, done: true };
      }
    };
  }
}

// 복잡한 비동기 패턴
class AsyncProcessor {
  async processItem(item: any): Promise<any> {
    await new Promise(resolve => setTimeout(resolve, 100));
    return item * 2;
  }

  async processBatch(items: any[]): Promise<any[]> {
    const promises = items.map(item => this.processItem(item));
    return Promise.all(promises);
  }

  async processWithRetry(
    item: any,
    maxRetries: number = 3
  ): Promise<any> {
    for (let attempt = 0; attempt < maxRetries; attempt++) {
      try {
        return await this.processItem(item);
      } catch (error) {
        if (attempt === maxRetries - 1) {
          throw error;
        }
        await new Promise(resolve =>
          setTimeout(resolve, 100 * (attempt + 1))
        );
      }
    }
    throw new Error("Max retries exceeded");
  }
}

// 이벤트 핸들러 패턴
type EventHandler<T = any> = (event: T) => void | Promise<void>;

interface EventEmitter<T extends Record<string, any>> {
  on<K extends keyof T>(event: K, handler: EventHandler<T[K]>): void;
  off<K extends keyof T>(event: K, handler: EventHandler<T[K]>): void;
  emit<K extends keyof T>(event: K, data: T[K]): void;
}

class SimpleEventEmitter<T extends Record<string, any>>
  implements EventEmitter<T>
{
  private handlers: Map<keyof T, EventHandler[]> = new Map();

  on<K extends keyof T>(event: K, handler: EventHandler<T[K]>): void {
    if (!this.handlers.has(event)) {
      this.handlers.set(event, []);
    }
    this.handlers.get(event)!.push(handler);
  }

  off<K extends keyof T>(event: K, handler: EventHandler<T[K]>): void {
    const handlers = this.handlers.get(event);
    if (handlers) {
      const index = handlers.indexOf(handler);
      if (index > -1) {
        handlers.splice(index, 1);
      }
    }
  }

  emit<K extends keyof T>(event: K, data: T[K]): void {
    const handlers = this.handlers.get(event);
    if (handlers) {
      handlers.forEach(handler => handler(data));
    }
  }
}

// 타입 가드
function isString(value: any): value is string {
  return typeof value === "string";
}

function isNumber(value: any): value is number {
  return typeof value === "number";
}

function isArray<T>(value: any): value is T[] {
  return Array.isArray(value);
}

function isValidUser(obj: any): obj is { name: string; age: number } {
  return (
    typeof obj === "object" &&
    obj !== null &&
    "name" in obj &&
    "age" in obj &&
    typeof obj.name === "string" &&
    typeof obj.age === "number"
  );
}

function processUserData(
  data: any
): { processed: boolean; user?: { name: string; age: number } } | null {
  if (isValidUser(data)) {
    return { processed: true, user: data };
  }
  return null;
}

// 복잡한 유니온 타입
type Status = "pending" | "processing" | "completed" | "failed" | "cancelled";

function getStatusColor(status: Status): string {
  const colors: Record<Status, string> = {
    pending: "yellow",
    processing: "blue",
    completed: "green",
    failed: "red",
    cancelled: "gray"
  };
  return colors[status];
}

// 복잡한 함수 시그니처
function complexFunctionSignature(
  required: string,
  optional?: number,
  ...rest: string[]
): {
  required: string;
  optional?: number;
  rest: string[];
} {
  return {
    required,
    optional,
    rest
  };
}

// 제네릭 제약 조건
interface HasId {
  id: string | number;
}

function findById<T extends HasId>(items: T[], id: string | number): T | undefined {
  return items.find(item => item.id === id);
}

// 매핑된 타입
type Readonly<T> = {
  readonly [P in keyof T]: T[P];
};

type Partial<T> = {
  [P in keyof T]?: T[P];
};

type Pick<T, K extends keyof T> = {
  [P in K]: T[P];
};

type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

// 조건부 타입을 사용한 유틸리티
type NonNullable<T> = T extends null | undefined ? never : T;

type KeysOfType<T, U> = {
  [K in keyof T]: T[K] extends U ? K : never;
}[keyof T];

// 복잡한 주석 패턴
class CommentedClass {
  /**
   * 매우 긴 독스트링을 가진 클래스.
   *
   * 이 클래스는 여러 줄에 걸친 설명을 가지고 있으며,
   * 복잡한 기능을 수행합니다.
   *
   * @param value - 초기값
   * @returns CommentedClass 인스턴스
   * @throws {Error} 잘못된 값이 전달될 경우
   *
   * @example
   * ```typescript
   * const obj = new CommentedClass(42);
   * console.log(obj.value); // 42
   * ```
   */
  constructor(public value: number) {
    // 인라인 주석
    this.value = value; // 값 설정
  }

  /**
   * 메서드 독스트링
   * @todo 구현 필요
   * @fixme 버그 수정 필요
   * @note 중요 참고사항
   */
  method(): void {
    // TODO: 구현 필요
    // FIXME: 버그 수정 필요
    // NOTE: 중요 참고사항
    // XXX: 위험한 코드
  }
}

// 믹스인 패턴
type Constructor<T = {}> = new (...args: any[]) => T;

function Timestamped<TBase extends Constructor>(Base: TBase) {
  return class extends Base {
    timestamp = new Date();
  };
}

function Activatable<TBase extends Constructor>(Base: TBase) {
  return class extends Base {
    isActive = false;

    activate() {
      this.isActive = true;
    }

    deactivate() {
      this.isActive = false;
    }
  };
}

class User {
  constructor(public name: string) {}
}

const TimestampedActivatableUser = Timestamped(Activatable(User));

// 프로퍼티 데코레이터
function readonly(target: any, propertyKey: string) {
  Object.defineProperty(target, propertyKey, {
    writable: false,
    configurable: false
  });
}

class ReadonlyClass {
  @readonly
  readonlyProperty: string = "cannot be changed";
}

// 파라미터 데코레이터
function validate(target: any, propertyKey: string, parameterIndex: number) {
  // 파라미터 검증 로직
}

class ValidatedClass {
  method(@validate value: number): void {
    // 메서드 구현
  }
}

// 복잡한 제네릭 함수
function identity<T>(arg: T): T {
  return arg;
}

function getProperty<T, K extends keyof T>(obj: T, key: K): T[K] {
  return obj[key];
}

function setProperty<T, K extends keyof T>(
  obj: T,
  key: K,
  value: T[K]
): void {
  obj[key] = value;
}

// 재귀적 타입
type JsonValue =
  | string
  | number
  | boolean
  | null
  | JsonValue[]
  | { [key: string]: JsonValue };

// 브랜드 타입
type Brand<T, B> = T & { __brand: B };

type UserId = Brand<string, "UserId">;
type ProductId = Brand<string, "ProductId">;

function createUserId(id: string): UserId {
  return id as UserId;
}

function createProductId(id: string): ProductId {
  return id as ProductId;
}

// 템플릿 리터럴 타입
type HttpMethod = "GET" | "POST" | "PUT" | "DELETE";
type ApiEndpoint = `/api/${string}`;
type FullEndpoint = `${HttpMethod} ${ApiEndpoint}`;

// 인덱스 시그니처
interface StringDictionary {
  [key: string]: string;
}

interface NumberDictionary {
  [key: number]: string;
}

// 복잡한 조건부 타입
type ExtractArrayType<T> = T extends (infer U)[] ? U : never;

type ExtractPromiseType<T> = T extends Promise<infer U> ? U : never;

type Flatten<T> = T extends (infer U)[] ? U : T;

// 유니온에서 교집합 추출
type UnionToIntersection<U> = (U extends any ? (x: U) => void : never) extends (
  x: infer I
) => void
  ? I
  : never;

